jni/src/common/Foreign/JNI/Unsafe.hs

```diff
@@ -1059,3 +1068,40 @@ getDirectBufferCapacity (upcast -> jbuffer) = do
       return capacity
     else
       throwIO DirectBufferFailed
+
+data GlobalRefCleaner o = GlobalRefCleaner (IORef [o]) (MVar Bool) (Async ())
+
+createGlobalRefCleaner :: Coercible o (J ty) => IO (GlobalRefCleaner o)
+createGlobalRefCleaner = do
+  wakeup <- newEmptyMVar
+  refs <- newIORef []
+  deletingThread <- async $ runInAttachedThread $
```

Comment:
I think you need to tack `runInBoundThread` here as `runInAttachedThread` does not ensure the thread is bound.